storage/innobase/lock/lock0lock.cc

```diff
@@ -4696,8 +4899,10 @@ static bool lock_release_on_prepare_try(trx_t *trx, bool unlock_unmodified)
     {
       ut_ad(!lock->index->table->is_temporary());
       bool supremum_bit= lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM);
+      /* if XA is being prepared, it must not own waiting locks */
+      ut_ad(!lock->is_waiting());
       bool rec_granted_exclusive_not_gap=
-        lock->is_rec_granted_exclusive_not_gap();
+        lock->is_rec_exclusive_not_gap();
```

Comment:
An assertion `!(type_mode & 2048U) || (type_mode & 512U)` is failing here during the execution of the test `rpl.rpl_parallel_xa_same_xid,mix`: https://buildbot.mariadb.org/#/builders/677/builds/12597
The `type_mode` is 2051, or `LOCK_INSERT_INTENTION | LOCK_S`. The assertion expression expects also `LOCK_GAP` to be set if `LOCK_INSERT_INTENTION` is set.